Kraken and Jormungandr : Change filling of "no_solution" error field

benoit-bst opened the pull request:
Add "no_solution" error field in jormungandr if journey no exist :

```json
"error": {
  "id": "no_solution",
  "message": "no solution found for this journey",
}
 ```

benoit-bst committed:
BugFix/NAVITIAII-2331: fill "no_solution" error field

Add "no_solution" error field in jormungandr if journey no exist :
"error": {
  "id": "no_solution"
  "message": "no solution found for this journey",
}

xlqian commented on source/jormungandr/jormungandr/interfaces/v1/Journeys.py:
```diff
@@ -529,10 +529,15 @@ def _set_specific_params(mod):
+            if not response.journeys:
                 logging.getLogger(__name__).debug("impossible to find journeys for the region {},"
                                                  " we'll try the next possible region ".format(r))
+                logging.getLogger(__name__).debug("Insert error field in response")
```
minor: i think you can merge the messages of L535 and L537

TeXitoi commented:
Tests are broken.

benoit-bst committed:
BugFix/NAVITIAII-2331 : Update to resolve broken tests

pbougue commented on source/routing/raptor_api.cpp:
```diff
@@ -662,8 +662,7 @@ static void add_pathes(PbCreator& pb_creator,
-static void
-make_pathes(PbCreator& pb_creator,
+static void make_pathes(PbCreator& pb_creator,
             const std::vector<navitia::routing::Path>& paths,
```
#VincentTatillon: better now, but weird indent :)

pbougue commented on source/routing/raptor_api.cpp:
```diff
@@ -675,7 +674,6 @@ make_pathes(PbCreator& pb_creator,
                origin, destination, datetimes, clockwise);
-
     if (pb_creator.empty_journeys()) {
         pb_creator.fill_pb_error(pbnavitia::Error::no_solution,
```
So you have to keep it here?
You couldn't update the tests so they match without that?

pbougue commented on source/jormungandr/jormungandr/interfaces/v1/Journeys.py:
```diff
@@ -529,17 +529,25 @@ def _set_specific_params(mod):
 
+            # If journeys list and error message is empty in response, we
+            # complete the error message fied
+            if not response.journeys and response.error.message == "":
```
It's more elegant to test `and not response.error.message`

pbougue reviewed (commented):
Is it a 200 http response in both cases?
It would be nice to add a test on the specific case of filtering all journeys, if it's "easily" doable.

benoit-bst committed:
BugFix/NAVITIAII-2331: Move the filling of error

- Refactoring of error handling in raptop_api/make_reponse
- Add comments

benoit-bst committed:
BugFix/NAVITIAII-2331: Add error message field (jor)

- If journeys list is empty and error field not exist, we create
  the error message field
- Bug fix in merge errors function

benoit-bst committed:
BugFix/NAVITIAII-2331: Add test unit test

xlqian commented:
you may want a rebase.... threre are a lot of irrelevant changes in your PR :)

xlqian commented on source/jormungandr/jormungandr/scenarios/new_default.py:
```diff
@@ -687,16 +687,18 @@ def merge_responses(responses):
+                # we need to merge the errors
+                merged_response.error.id = response_pb2.Error.no_solution
+                merged_response.error.message = "several errors occured: \n * {}"\
+                    .format("\n * ".join([m.message for m in errors.values()]))
```
the whole bloc can be written as follows:

```python
if len(errors) == 1:
    merged_response.error.id = next(errors.values()).id   # it seems values() returns a generator, not sure, to be tested
    merged_response.error.id = response_pb2.Error.no_solution
    merged_response.error.message = next(errors.values()).message
elif len(errors) > 1:
    merged_response.error.id = response_pb2.Error.no_solution
    merged_response.error.message = "several errors occured: \n * {}"\
        .format("\n * ".join((m.message for m in errors.values())))
```

benoit-bst replied:
Yes, it is a better syntax

xlqian commented on source/jormungandr/tests/journey_common_tests.py:
```diff
@@ -853,6 +853,30 @@ def test_journey_from_non_valid_object_type(self):
+            with "no_solution" id and "no solution found for this journey"
+            message.
+            """
+            query = "journeys?from={from_sp}&to={to_sp}&datetime={datetime}&_override_scenario=new_default" \
```
i'm wondering if `_override_scenario=new_default` is necessary

benoit-bst replied:
True :-)

benoit-bst committed:
BugFix/NAVITIAII-2331: Suppress flag in test

Suppress _override_scenario=new_default

benoit-bst committed:
BugFix/NAVITIAII-2331: Improve syntax

xlqian reviewed (approved):
:confetti_ball:

pbougue commented on source/jormungandr/jormungandr/interfaces/v1/Journeys.py:
```diff
@@ -529,17 +529,25 @@ def _set_specific_params(mod):
+                logging.getLogger(__name__).debug("impossible to find journeys for the region {},"
+                                                 " insert error field in response ".format(r))
+                response.error.id = response_pb2.Error.no_solution
+                response.error.message = "no solution found for this journey"
```
Don't know if managing the status also can be usefull (since it's managed all in one place in kraken, maybe also in jormun)

benoit-bst replied:
It seems more coherent. I add it

pbougue commented on source/jormungandr/jormungandr/scenarios/new_default.py:
```diff
@@ -687,13 +687,14 @@ def merge_responses(responses):
-            merged_response.error.message = list(errors.values())[0].message
-        else:
-            # we need to merge the errors
+            merged_response.error.id = errors.values()[0].id
```
not sure about this `list` removal, it was added specifically by @kinnou02 in #2126

pbougue commented on source/jormungandr/tests/routing_tests_new_default.py:
```diff
@@ -246,6 +246,9 @@ class TestNewDefaultJourneysWithPtref(JourneysWithPtref, NewDefaultScenarioAbstr
     pass
 
+@config({"scenario": "new_default"})
+class TestNewDefaultJourneysNoRegion(AddErrorFieldInJormun, NewDefaultScenarioAbstractTestFixture):
```
2 classes with the same name : doesn't sound good

pbougue commented on source/routing/raptor_api.cpp:
```diff
@@ -1058,97 +1051,131 @@ void make_response(navitia::PbCreator& pb_creator,
+        make_pathes(pb_creator,
+                    pathes,
+                    worker,
+                    get_direct_path(worker, origin, destination),
```
you already computed it, and have it available in `direct_path`

benoit-bst committed:
Fill Response Type in Jormun

We Fill Response Type field in Jormun
when we add Error field when journeys
not exist

benoit-bst committed:
Add Python 3 compatibility

benoit-bst committed:
fix protobuf enum

benoit-bst committed:
Change test name

benoit-bst committed:
Avoid a function call (direct path)

kinnou02 commented on source/routing/raptor_api.cpp:
```diff
@@ -1058,97 +1051,131 @@ void make_response(navitia::PbCreator& pb_creator,
-        return;
     }
+    // classical case : We have departures and destinations
+    else {
```
I don't like having a `else` that isn't on the same line than the closing bracket

kinnou02 reviewed (commented):
There is a lot of refactoring in this PR, it should be done in another PR to make reviewing easier.
Or at least in different commitsif the refactoring is "required" by the change made.

benoit-bst commented:
https://jira.canaltp.fr/browse/NAVITIAII-2331
